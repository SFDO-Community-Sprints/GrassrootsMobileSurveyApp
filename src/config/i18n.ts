import * as Localization from 'expo-localization';
import i18n from 'i18n-js';

import ne from './locale/ne_IN.json';
import en from './locale/en_US.json';

import { DEFAULT_SF_LANGUAGE } from '../constants';
import { toShortLocale } from '../utility';

//expo-localization gives the result in a language-region format
const deviceLocale: string =
  Localization.locale.indexOf('-') !== 0 ? Localization.locale.split('-')[0] : Localization.locale;

i18n.defaultLocale = deviceLocale === 'ne' ? deviceLocale : toShortLocale(DEFAULT_SF_LANGUAGE.code);
i18n.locale = deviceLocale === 'ne' ? deviceLocale : toShortLocale(DEFAULT_SF_LANGUAGE.code);
i18n.fallbacks = true;
i18n.translations = { en, ne };

export default i18n;
