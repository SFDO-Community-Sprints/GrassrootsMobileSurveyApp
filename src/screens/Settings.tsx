import React, { useEffect, useState } from 'react';
import { FlatList, ImageBackground } from 'react-native';
import { Card, Icon, Divider, ListItem, Text } from '@rneui/themed';
import NetInfo from '@react-native-community/netinfo';
import * as Application from 'expo-application';
import * as SecureStore from 'expo-secure-store';

import { useLocalizationContext } from '../context/localizationContext';
import { retrieveAllMetadata } from '../services/describe';
import { forceLogout } from '../services/session';
import { Loader } from '../components';

import {
  APP_THEME,
  BACKGROUND_IMAGE_SOURCE,
  BACKGROUND_STYLE,
  BACKGROUND_IMAGE_STYLE,
  APP_FONTS,
  SECURE_STORE_KEYS,
  ASYNC_STORAGE_KEYS,
  DEFAULT_SF_LANGUAGE,
} from '../constants';
import { logger } from '../utility/logger';
import { toShortLocale } from '../utility';
import { notifySuccess, notifyError } from '../utility/notification';
import { storeOnlineSurveys } from '../services/salesforce/survey';
import { buildDictionary } from '../services/dictionary';
import { hasUnsyncedSurveys } from '../services/database/localSurvey';
import { storeContacts } from '../services/salesforce/contact';
import { useAuthContext } from '../context/authContext';
import { Language } from '../types/metadata';

export default function Settings() {
  const [email, setEmail] = useState('');
  const [showsSpinner, setShowsSpinner] = useState(false);
  const { t, locale, setLocale } = useLocalizationContext();
  const authContext = useAuthContext();

  const [languages, setLanguages] = useState<Array<Language>>([DEFAULT_SF_LANGUAGE]);

  useEffect(() => {
    const prepare = async () => {
      const email = await SecureStore.getItemAsync(SECURE_STORE_KEYS.EMAIL);
      const languages = await storage.load({ key: ASYNC_STORAGE_KEYS.LANGUAGES });
      setLanguages(languages);
      if (email) {
        setEmail(email);
      }
    };
    prepare();
  }, []);

  const renderItem = ({ item }) => {
    return (
      <ListItem
        onPress={() => {
          setLocale(toShortLocale(item.code));
          logger('FINE', 'Settings', `change locale to: ${item.code}`);
        }}
      >
        {toShortLocale(item.code) === locale && <Icon name="check" size={20} color={APP_THEME.APP_BASE_COLOR} />}
        <ListItem.Content>
          <ListItem.Title style={{ fontFamily: APP_FONTS.FONT_REGULAR }}>{item.name}</ListItem.Title>
        </ListItem.Content>
      </ListItem>
    );
  };

  const isInternetReachable = async () => {
    const netInfo = await NetInfo.fetch();
    if (!netInfo.isInternetReachable) {
      notifyError('No network connection. Confirm your network connectivity and try again.');
      return false;
    }
    return true;
  };

  return (
    <ImageBackground source={BACKGROUND_IMAGE_SOURCE} style={BACKGROUND_STYLE} imageStyle={BACKGROUND_IMAGE_STYLE}>
      <Loader loading={showsSpinner} />
      <Card>
        <Card.Title style={{ fontFamily: APP_FONTS.FONT_BOLD }}>{t('LANGUAGE')}</Card.Title>
        <FlatList
          data={languages}
          renderItem={renderItem}
          keyExtractor={(item, index) => {
            return index.toString();
          }}
          ItemSeparatorComponent={() => {
            return <Divider style={{ backgroundColor: APP_THEME.APP_BORDER_COLOR }} />;
          }}
        />
      </Card>
      <Card>
        <Card.Title style={{ fontFamily: APP_FONTS.FONT_BOLD }}>{t('INFORMATION')}</Card.Title>
        <Divider style={{ backgroundColor: APP_THEME.APP_BORDER_COLOR }} />
        <Text style={{ textAlign: 'center', paddingTop: 15, paddingBottom: 15 }}>
          {t('VERSION')} {Application.nativeApplicationVersion}
        </Text>
        <Divider style={{ backgroundColor: APP_THEME.APP_BORDER_COLOR }} />
        <Text style={{ textAlign: 'center', paddingTop: 15 }}>
          {t('LOGGING_IN_AS')} {email}
        </Text>
      </Card>
      <Card>
        <Card.Title style={{ fontFamily: APP_FONTS.FONT_BOLD }}>{t('MAINTENANCE')}</Card.Title>
        <ListItem
          onPress={async () => {
            if (!(await isInternetReachable())) {
              return;
            }
            if (await hasUnsyncedSurveys()) {
              notifyError('You cannot reload metadata and surveys until all the surveys are synced.');
              return;
            }
            setShowsSpinner(true);
            try {
              await storeContacts();
              await retrieveAllMetadata();
              await storeOnlineSurveys();
              await buildDictionary();
              notifySuccess('Successfully refreshed metadata.');
            } catch (e) {
              notifyError('Unexpected error occcured while refreshing. Contact your administrator and login again.');
              await forceLogout(authContext);
            } finally {
              setShowsSpinner(false);
            }
          }}
          topDivider
        >
          <Icon name="cloud-download" color={APP_THEME.APP_LIGHT_FONT_COLOR} />
          <ListItem.Content>
            <ListItem.Title style={{ fontFamily: APP_FONTS.FONT_REGULAR }}>Reload Metadata and Survey</ListItem.Title>
          </ListItem.Content>
        </ListItem>
      </Card>
    </ImageBackground>
  );
}
