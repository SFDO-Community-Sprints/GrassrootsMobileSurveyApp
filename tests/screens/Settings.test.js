import React from 'react';
import { render } from '@testing-library/react-native';

import i18n from '../../src/config/i18n';
import LocalizationContext from '../../src/context/localizationContext';
import Settings from '../../src/screens/Settings';

describe('<Settings />', () => {
  it('render', () => {
    const locale = i18n.locale;
    const localizationContext = {
      t: jest.fn(),
      locale,
      setLocale: jest.fn(),
    };

    const screen = render(
      <LocalizationContext.Provider value={localizationContext}>
        <Settings />
      </LocalizationContext.Provider>
    );
    expect(screen.toJSON()).toMatchSnapshot();
  });
});
