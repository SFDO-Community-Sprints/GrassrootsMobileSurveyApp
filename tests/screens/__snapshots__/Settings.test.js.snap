// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Settings /> render 1`] = `
<View
  accessibilityIgnoresInvertColors={true}
  style={
    Object {
      "backgroundColor": "white",
      "flex": 1,
    }
  }
>
  <Image
    source={1}
    style={
      Array [
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        Object {
          "height": undefined,
          "width": undefined,
        },
        Object {
          "resizeMode": "contain",
          "top": 857.7142857142858,
        },
      ]
    }
  />
  <Modal
    hardwareAccelerated={false}
    transparent={true}
    visible={false}
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "white",
          "flex": 1,
          "justifyContent": "center",
          "opacity": 0.8,
        }
      }
    >
      <ActivityIndicator
        animating={false}
        color="#01579b"
        hidesWhenStopped={true}
        size="large"
      />
      <Text
        style={
          Object {
            "color": "#01579b",
            "fontFamily": "SalesforceSans-Regular",
            "fontSize": 36,
            "paddingTop": 10,
          }
        }
      >
         
         
      </Text>
    </View>
  </Modal>
  <View
    style={
      Object {
        "backgroundColor": "#ffffff",
        "borderColor": "#e1e8ee",
        "borderWidth": 1,
        "margin": 15,
        "marginBottom": 0,
        "padding": 15,
        "shadowColor": "rgba(0,0,0, .2)",
        "shadowOffset": Object {
          "height": 0,
          "width": 0,
        },
        "shadowOpacity": 1,
        "shadowRadius": 1,
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "transparent",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#43484d",
            "fontFamily": "SalesforceSans-Bold",
            "fontSize": 17.5,
            "fontWeight": "bold",
            "marginBottom": 15,
            "textAlign": "center",
          }
        }
        testID="cardTitle"
      >
        
      </Text>
      <RCTScrollView
        ItemSeparatorComponent={[Function]}
        data={
          Array [
            Object {
              "code": "en",
              "name": "English",
            },
            Object {
              "code": "ne",
              "name": "Nepali",
            },
          ]
        }
        disableVirtualization={false}
        getItem={[Function]}
        getItemCount={[Function]}
        horizontal={false}
        initialNumToRender={10}
        keyExtractor={[Function]}
        maxToRenderPerBatch={10}
        onContentSizeChange={[Function]}
        onEndReachedThreshold={2}
        onLayout={[Function]}
        onMomentumScrollEnd={[Function]}
        onScroll={[Function]}
        onScrollBeginDrag={[Function]}
        onScrollEndDrag={[Function]}
        removeClippedSubviews={false}
        renderItem={[Function]}
        scrollEventThrottle={50}
        stickyHeaderIndices={Array []}
        updateCellsBatchingPeriod={50}
        viewabilityConfigCallbackPairs={Array []}
        windowSize={21}
      >
        <View>
          <View
            onLayout={[Function]}
            style={null}
          >
            <View
              accessible={true}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "#ffffff",
                    "borderColor": "#bcbbc1",
                    "flexDirection": "row",
                    "padding": 14,
                  }
                }
                testID="padView"
              >
                <View
                  color="#01579b"
                  name="check"
                  size={20}
                />
                <View
                  style={
                    Object {
                      "paddingLeft": 16,
                    }
                  }
                />
                <View
                  style={
                    Object {
                      "alignItems": "flex-start",
                      "flex": 1,
                      "justifyContent": "center",
                    }
                  }
                  theme={
                    Object {
                      "colors": Object {
                        "black": "#242424",
                        "disabled": "hsl(208, 8%, 90%)",
                        "divider": "#bcbbc1",
                        "error": "#ff190c",
                        "grey0": "#393e42",
                        "grey1": "#43484d",
                        "grey2": "#5e6977",
                        "grey3": "#86939e",
                        "grey4": "#bdc6cf",
                        "grey5": "#e1e8ee",
                        "greyOutline": "#bbb",
                        "platform": Object {
                          "android": Object {
                            "error": "#f44336",
                            "grey": "rgba(0, 0, 0, 0.54)",
                            "primary": "#2196f3",
                            "secondary": "#9C27B0",
                            "success": "#4caf50",
                            "warning": "#ffeb3b",
                          },
                          "ios": Object {
                            "error": "#ff3b30",
                            "grey": "#7d7d7d",
                            "primary": "#007aff",
                            "searchBg": "#dcdce1",
                            "secondary": "#5856d6",
                            "success": "#4cd964",
                            "warning": "#ffcc00",
                          },
                        },
                        "primary": "#2089dc",
                        "searchBg": "#303337",
                        "secondary": "#ca71eb",
                        "success": "#52c41a",
                        "warning": "#faad14",
                        "white": "#ffffff",
                      },
                    }
                  }
                >
                  <Text
                    style={
                      Object {
                        "backgroundColor": "transparent",
                        "color": "#242424",
                        "fontFamily": "SalesforceSans-Regular",
                        "fontSize": 17,
                      }
                    }
                    testID="listItemTitle"
                  >
                    English
                  </Text>
                </View>
              </View>
            </View>
            <View
              style={
                Object {
                  "backgroundColor": "#dedcdb",
                  "height": 0.5,
                }
              }
            />
          </View>
          <View
            onLayout={[Function]}
            style={null}
          >
            <View
              accessible={true}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "#ffffff",
                    "borderColor": "#bcbbc1",
                    "flexDirection": "row",
                    "padding": 14,
                  }
                }
                testID="padView"
              >
                <View
                  style={
                    Object {
                      "alignItems": "flex-start",
                      "flex": 1,
                      "justifyContent": "center",
                    }
                  }
                  theme={
                    Object {
                      "colors": Object {
                        "black": "#242424",
                        "disabled": "hsl(208, 8%, 90%)",
                        "divider": "#bcbbc1",
                        "error": "#ff190c",
                        "grey0": "#393e42",
                        "grey1": "#43484d",
                        "grey2": "#5e6977",
                        "grey3": "#86939e",
                        "grey4": "#bdc6cf",
                        "grey5": "#e1e8ee",
                        "greyOutline": "#bbb",
                        "platform": Object {
                          "android": Object {
                            "error": "#f44336",
                            "grey": "rgba(0, 0, 0, 0.54)",
                            "primary": "#2196f3",
                            "secondary": "#9C27B0",
                            "success": "#4caf50",
                            "warning": "#ffeb3b",
                          },
                          "ios": Object {
                            "error": "#ff3b30",
                            "grey": "#7d7d7d",
                            "primary": "#007aff",
                            "searchBg": "#dcdce1",
                            "secondary": "#5856d6",
                            "success": "#4cd964",
                            "warning": "#ffcc00",
                          },
                        },
                        "primary": "#2089dc",
                        "searchBg": "#303337",
                        "secondary": "#ca71eb",
                        "success": "#52c41a",
                        "warning": "#faad14",
                        "white": "#ffffff",
                      },
                    }
                  }
                >
                  <Text
                    style={
                      Object {
                        "backgroundColor": "transparent",
                        "color": "#242424",
                        "fontFamily": "SalesforceSans-Regular",
                        "fontSize": 17,
                      }
                    }
                    testID="listItemTitle"
                  >
                    Nepali
                  </Text>
                </View>
              </View>
            </View>
          </View>
        </View>
      </RCTScrollView>
    </View>
  </View>
  <View
    style={
      Object {
        "backgroundColor": "#ffffff",
        "borderColor": "#e1e8ee",
        "borderWidth": 1,
        "margin": 15,
        "marginBottom": 0,
        "padding": 15,
        "shadowColor": "rgba(0,0,0, .2)",
        "shadowOffset": Object {
          "height": 0,
          "width": 0,
        },
        "shadowOpacity": 1,
        "shadowRadius": 1,
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "transparent",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#43484d",
            "fontFamily": "SalesforceSans-Bold",
            "fontSize": 17.5,
            "fontWeight": "bold",
            "marginBottom": 15,
            "textAlign": "center",
          }
        }
        testID="cardTitle"
      >
        
      </Text>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#ffffff",
              "borderColor": "#bcbbc1",
              "borderTopWidth": 0.5,
              "flexDirection": "row",
              "padding": 14,
            }
          }
          testID="padView"
        >
          <View
            color="#706e6b"
            name="cloud-download"
          />
          <View
            style={
              Object {
                "paddingLeft": 16,
              }
            }
          />
          <View
            style={
              Object {
                "alignItems": "flex-start",
                "flex": 1,
                "justifyContent": "center",
              }
            }
            theme={
              Object {
                "colors": Object {
                  "black": "#242424",
                  "disabled": "hsl(208, 8%, 90%)",
                  "divider": "#bcbbc1",
                  "error": "#ff190c",
                  "grey0": "#393e42",
                  "grey1": "#43484d",
                  "grey2": "#5e6977",
                  "grey3": "#86939e",
                  "grey4": "#bdc6cf",
                  "grey5": "#e1e8ee",
                  "greyOutline": "#bbb",
                  "platform": Object {
                    "android": Object {
                      "error": "#f44336",
                      "grey": "rgba(0, 0, 0, 0.54)",
                      "primary": "#2196f3",
                      "secondary": "#9C27B0",
                      "success": "#4caf50",
                      "warning": "#ffeb3b",
                    },
                    "ios": Object {
                      "error": "#ff3b30",
                      "grey": "#7d7d7d",
                      "primary": "#007aff",
                      "searchBg": "#dcdce1",
                      "secondary": "#5856d6",
                      "success": "#4cd964",
                      "warning": "#ffcc00",
                    },
                  },
                  "primary": "#2089dc",
                  "searchBg": "#303337",
                  "secondary": "#ca71eb",
                  "success": "#52c41a",
                  "warning": "#faad14",
                  "white": "#ffffff",
                },
              }
            }
          >
            <Text
              style={
                Object {
                  "backgroundColor": "transparent",
                  "color": "#242424",
                  "fontFamily": "SalesforceSans-Regular",
                  "fontSize": 17,
                }
              }
              testID="listItemTitle"
            >
              Reload Metadata and Survey
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
